docs/rfcs/042-compute-pageserver-communicator.md

```diff
@@ -0,0 +1,197 @@
+# Compute <-> Pageserver Communicator Rewrite
+
+## Summary
+
+## Motivation
+
+- prefetching logic is complicated
+- handling async libpq connections in C code are difficult and error-prone
+- only few people are comfortable working on the code
+- new AIO (maybe) coming up in Postgres v18
+- cannot process prefetch replies until another I/O function is called. Makes it impossible to accurately measure when a reply was received
+- every backend opens a separate connection to pageservers -> lots of connections, first query in backend is slow
+
+- desire for better protocol, not libpq-based
+
+- By writing the "communicator" as a separate rust module, it can be reused in
+  tests, outside PostgreSQL.
+
+## Non Goals (if relevant)
+
+- We will keep LFC unmodified for now. It might be a good idea to rewrite it
+  too, but it's out of scope here.
+
+- We should consider a similar rewrite for the walproposer - safekeeper
+  communication, but it's out of scope for this RFC
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+
+- Most changes are to the neon extension in compute.
+
+- Pageserver, to implement the new protocol.
+
+## Proposed implementation
+
+- we will use the new implementation with all PostgreSQL versions.
+- we will have a feature flag to switch between old and new communicator. Once we're
+  comfortable with the new communicator, remove old code and protocol.
+
+- What about relation size cache? Out of scope? Or move it to the communicator process,
+  and have smgrnblocks() requests always through communicator process?
+
+### Communicator process
+
+There is one communicator process in the Postgres server. It's a background
+worker process. It handles all communication with the pageservers.
+
+The communicator process is written in a mix of C and Rust. Mostly in Rust, but
```

Comment:
I are absolutely sure that mixing C and Rust code in the same process is good idea?
Yes, Rust is "safe", but it needs to access data in shared memory. Not sure if it can be done in "safe" way. 
Certainly rust allows use to use the same libraries at compute/PS side, for example serve for request serialization/deserializaiton. But still not sure whether there are more pros than cons.

Reply:
several mentioned interfaces are already done:

- logs and error reporting: https://docs.rs/pgrx/latest/pgrx/#macros
- shmem: https://docs.rs/pgrx/latest/pgrx/shmem/trait.PgSharedMemoryInitialization.html

Reply:
May be I am missing something, but it is just shmem initialization, not shmem access.